File: backend/app/pipeline/classify.py
"""Contract type and clause theme classification (keyword heuristics v0)."""

from __future__ import annotations

import re
from typing import Literal

ContractType = Literal["lease", "internship_offer", "freelance", "unknown"]
Theme = Literal[
    "payment",
    "termination",
    "ip",
    "confidentiality",
    "disputes",
    "renewal",
    "deposit",
    "maintenance",
    "subletting",
    "scope",
    "indemnity",
    "general",
]


TYPE_KEYWORDS: dict[ContractType, list[str]] = {
    "lease": ["landlord", "tenant", "rent", "security deposit", "premises", "lease term"],
    "internship_offer": ["intern", "internship", "at-will", "employment", "offer letter", "wage", "salary"],
    "freelance": ["independent contractor", "sow", "statement of work", "milestone", "invoice", "services"],
    "unknown": [],
}


THEME_PATTERNS: list[tuple[Theme, re.Pattern[str]]] = [
    ("payment", re.compile(r"payment|compensation|invoice|fee|rent|deposit|salary|wage", re.I)),
    ("termination", re.compile(r"terminat|cancel|breach|notice period|at-will", re.I)),
    ("ip", re.compile(r"intellectual property|invention|work product|assignment|derivative", re.I)),
    ("confidentiality", re.compile(r"confidential|non-disclosure|proprietary|trade secret", re.I)),
    ("disputes", re.compile(r"arbitrat|mediation|jurisdiction|governing law|class action", re.I)),
    ("renewal", re.compile(r"renew|automatically renew|month-to-month|extension", re.I)),
    ("deposit", re.compile(r"security deposit|damage deposit|withhold", re.I)),
    ("maintenance", re.compile(r"repair|maintain|habitable|landlord|tenant obligation", re.I)),
    ("subletting", re.compile(r"sublet|assign|transfer|guest", re.I)),
    ("scope", re.compile(r"scope of services|deliverable|change order|milestone", re.I)),
    ("indemnity", re.compile(r"indemnif|hold harmless|liability cap", re.I)),
]


def detect_contract_type(text: str, hint: str | None) -> ContractType:
    if hint and hint in ("lease", "internship_offer", "freelance"):
        return hint  # type: ignore[return-value]
    if hint == "auto" or not hint:
        scores: dict[ContractType, int] = {"lease": 0, "internship_offer": 0, "freelance": 0, "unknown": 0}
        low = text.lower()
        for ctype, kws in TYPE_KEYWORDS.items():
            if ctype == "unknown":
                continue
            scores[ctype] = sum(1 for k in kws if k in low)
        best = max(scores, key=lambda k: scores[k])
        if scores[best] >= 2:
            return best
        return "unknown"
    return "unknown"


def classify_theme(clause_text: str) -> Theme:
    for theme, pat in THEME_PATTERNS:
        if pat.search(clause_text):
            return theme
    return "general"


def risk_level_for_clause(theme: Theme, clause_text: str, contract_type: ContractType) -> str:
    low = clause_text.lower()
    aggressive = any(
        x in low
        for x in [
            "sole discretion",
            "unlimited",
            "irrevocable",
            "perpetual",
            "automatic renewal",
            "binding arbitration",
            "exclusive",
            "assigns all",
        ]
    )
    vague = any(x in low for x in ["reasonable", "as determined", "may", "sole"]) and len(clause_text) > 200

    if aggressive:
        return "high"
    if vague or theme in ("ip", "indemnity", "disputes"):
        return "medium"
    return "low"

File: ios/CoParse/Engine/Classify.swift
import Foundation

enum ContractType: String, CaseIterable, Identifiable {
    case lease
    case internshipOffer = "internship_offer"
    case freelance
    case unknown

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .lease: return "Lease"
        case .internshipOffer: return "Internship / offer"
        case .freelance: return "Freelance"
        case .unknown: return "Unknown"
        }
    }

    var defaultRole: String {
        switch self {
        case .lease: return "renter"
        case .internshipOffer: return "student_intern"
        case .freelance: return "freelancer"
        case .unknown: return "renter"
        }
    }
}

enum ClauseTheme: String {
    case payment, termination, ip, confidentiality, disputes, renewal
    case deposit, maintenance, subletting, scope, indemnity, general
}

enum AnalysisClassify {
    static let typeKeywords: [ContractType: [String]] = [
        .lease: [
            "landlord", "tenant", "rent", "security deposit", "premises", "lease term",
            "lessee", "lessor", "apartment", "dwelling", "move-in", "occupancy",
        ],
        .internshipOffer: [
            "intern", "internship", "at-will", "employment", "offer letter", "wage", "salary",
            "stipend", "supervisor", "orientation", "benefits", "exempt", "non-exempt",
        ],
        .freelance: [
            "independent contractor", "sow", "statement of work", "milestone", "invoice", "services",
            "deliverable", "client", "contractor", "retainer", "change order", "work for hire",
        ],
        .unknown: [],
    ]

    static let themePatterns: [(ClauseTheme, NSRegularExpression)] = {
        let specs: [(ClauseTheme, String)] = [
            (.payment, #"payment|compensation|invoice|fee|rent|deposit|salary|wage|stipend|late fee|net\s*\d+"#),
            (.termination, #"terminat|cancel|breach|notice period|at-will|severance|resignation"#),
            (.ip, #"intellectual property|invention|work product|assignment|derivative|work for hire|work-for-hire"#),
            (.confidentiality, #"confidential|non-disclosure|proprietary|trade secret|nda\b"#),
            (.disputes, #"arbitrat|mediation|jurisdiction|governing law|class action|venue|attorney.?s fees"#),
            (.renewal, #"renew|automatically renew|month-to-month|extension|evergreen"#),
            (.deposit, #"security deposit|damage deposit|withhold|deposit return"#),
            (.maintenance, #"repair|maintain|habitable|landlord|tenant obligation|utilities"#),
            (.subletting, #"sublet|assign|transfer|guest|roommate"#),
            (.scope, #"scope of services|deliverable|change order|milestone|revision"#),
            (.indemnity, #"indemnif|hold harmless|liability cap|limitation of liability|consequential damages"#),
        ]
        return specs.compactMap { theme, pattern in
            guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
            return (theme, re)
        }
    }()

    static func detectContractType(text: String, hint: String?) -> ContractType {
        if let hint, let typed = ContractType(rawValue: hint), typed != .unknown {
            return typed
        }
        if hint == nil || hint == "auto" || hint == "unknown" {
            let low = text.lowercased()
            var scores: [ContractType: Int] = [.lease: 0, .internshipOffer: 0, .freelance: 0]
            for (ctype, kws) in typeKeywords where ctype != .unknown {
                scores[ctype] = kws.reduce(0) { $0 + (low.contains($1) ? 1 : 0) }
            }
            let best = scores.max(by: { $0.value < $1.value })
            if let best, best.value >= 2 { return best.key }
            return .unknown
        }
        return .unknown
    }

    static func classifyTheme(_ clauseText: String) -> ClauseTheme {
        let range = NSRange(clauseText.startIndex..<clauseText.endIndex, in: clauseText)
        for (theme, re) in themePatterns {
            if re.firstMatch(in: clauseText, options: [], range: range) != nil {
                return theme
            }
        }
        return .general
    }

    static func riskLevel(theme: ClauseTheme, clauseText: String, contractType: ContractType) -> String {
        let low = clauseText.lowercased()
        let aggressive = [
            "sole discretion", "unlimited", "irrevocable", "perpetual", "automatic renewal",
            "binding arbitration", "exclusive", "assigns all", "waive all", "without notice",
            "liquidated damages", "non-compete", "noncompete", "injunctive relief",
        ].contains { low.contains($0) }
        let leaseRed = contractType == .lease && [
            "no pets", "tenant responsible for all repairs", "as-is", "landlord may enter at any time",
        ].contains { low.contains($0) }
        let vague = (["reasonable", "as determined", "may", "sole"].contains { low.contains($0) }) && clauseText.count > 200

        if aggressive || leaseRed { return "high" }
        if vague || theme == .ip || theme == .indemnity || theme == .disputes { return "medium" }
        if theme == .payment || theme == .termination || theme == .deposit { return "medium" }
        return "low"
    }
}
